src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java

```diff
@@ -389,41 +315,209 @@ private static <T extends Enum<T>> int mappedEnumLookup(T value, MethodHandles.L
+                    result = SENTINEL;
+                }
+
+                resolvedEnum[labelIndex] = result;
+            }
+
+            return result;
+        }
     }
 
     private static final class EnumMap {
         @Stable
         public int[] map;
     }
+
+    /*
+     * Construct test chains for labels inside switch, to handle switch repeats:
+     * switch (idx) {
+     *     case 0 -> if (selector matches label[0]) return 0;
+     *     case 1 -> if (selector matches label[1]) return 1;
+     *     ...
+     * }
+     */
+    @SuppressWarnings("removal")
+    private static MethodHandle generateInnerClass(MethodHandles.Lookup caller, Object[] labels) {
+        List<EnumDesc<?>> enumDescs = new ArrayList<>();
+
+        byte[] classBytes = Classfile.of().build(ClassDesc.of(typeSwitchClassName(caller.lookupClass())), clb -> {
+            clb.withFlags(AccessFlag.FINAL, AccessFlag.SYNTHETIC)
+              .withMethod("typeSwitch", MethodTypeDesc.ofDescriptor("(Ljava/lang/Object;ILjava/util/function/BiFunction;)I"), Classfile.ACC_STATIC, mb -> {
+                  mb.withFlags(AccessFlag.PUBLIC, AccessFlag.FINAL, AccessFlag.STATIC)
+                    .withCode(cb -> {
+                        cb.aload(0);
+                        Label nonNullLabel = cb.newLabel();
+                        cb.if_nonnull(nonNullLabel);
+                        cb.iconst_m1();
+                        cb.ireturn();
+                        cb.labelBinding(nonNullLabel);
+                        if (labels.length == 0) {
+                            cb.constantInstruction(0)
+                              .ireturn();
+                        }
+                        cb.iload(1);
+                        Label dflt = cb.newLabel();
+                        record Element(Label target, Label next, Object label) {}
+                        List<Element> cases = new ArrayList<>();
+                        List<SwitchCase> switchCases = new ArrayList<>();
+                        Object lastLabel = null;
+                        for (int idx = labels.length - 1; idx >= 0; idx--) {
+                            Object currentLabel = labels[idx];
+                            Label target = cb.newLabel();
+                            Label next;
+                            if (lastLabel == null) {
+                                next = dflt;
+                            } else if (lastLabel.equals(currentLabel)) {
+                                next = cases.getLast().next();
+                            } else {
+                                next = cases.getLast().target();
+                            }
+                            lastLabel = currentLabel;
+                            cases.add(new Element(target, next, currentLabel));
+                            switchCases.add(SwitchCase.of(idx, target));
+                        }
+                        cases = cases.reversed();
```

Comment:
switchCases for tableswitch do not need to be pre-ordered, code builder does the processign

Reply:
I guess I would prefer to keep the reverse here, to reduce the cognitive load, as `cases` must be processed in original order (and hence reversed here), and it may not be clear why reverse one, and not the other. (Given the `List` is an `ArrayList`, the reverse operation should not have much impact on anything, if I read how it is implemented correctly.)

Reply:
I think you don't need to reverse any of the lists.